AVRO-2001 Adding support for doc attribute [ruby]

lostphilosopher opened the pull request:
**AVRO-2001**: https://issues.apache.org/jira/browse/AVRO-2001

The Avro schema and protocol specification(http://avro.apache.org/docs/current/spec.html) allows for an optional attribute `doc` defined as, "a JSON string describing this field for users." Users should be able to include a doc with each field for `records` (http://avro.apache.org/docs/current/spec.html#N10085), `record fields`, and `enums` (http://avro.apache.org/docs/current/spec.html#N10186). The `doc` attribute is also in the spec for `protocols` and `protocol messages` (http://avro.apache.org/docs/1.8.1/spec.html#N105C6 and http://avro.apache.org/docs/1.8.1/spec.html#N105F1).

This doc attribute should be stripped when transforming into "Parsing Canonical Form" (http://avro.apache.org/docs/current/spec.html#N10815).

Currently, the Ruby Avro library (https://github.com/apache/avro/tree/master/lang/ruby) does not support `doc`. It is supported in other languages libraries, for example PHP (https://github.com/apache/avro/blob/master/lang/php/lib/avro/schema.php#L46).

tjwp commented on lang/ruby/test/test_protocol.rb:
```diff
@@ -196,4 +215,9 @@ def test_namespaces
+
+  def test_support_for_doc_attribute
+    original = Protocol.parse(EXAMPLES.last.protocol_string)
+    assert_equal original.types.first.fields.first.doc, 'documentation'
```
This isn't testing the doc attribute on a Message, which is added implemented in this PR, just doc on a Field.

lostphilosopher commented:
Solid feedback on all counts. I'll have another draft soon.

lostphilosopher commented on lang/ruby/test/test_schema.rb:
```diff
@@ -160,4 +160,101 @@ def test_to_avro_handles_falsey_defaults
     }
   end
+
+  def test_support_for_doc_attribute
```
I have an RSpec background. I don't have `test-unit` experience. Should these all be separate tests? How would you structure these?

lostphilosopher commented:
@tjwp This is ready for another look. I've added `doc` support to the classes you listed, and fixed my wrong headed unit test. 

Right now I've just added two unit tests, each with multiple asserts for testing `doc` support in various contexts. Perhaps they should be separate tests? I'm happy to break them down or do whatever you recommend. I'm new to `test-unit` there may be syntax I'm not aware of. I could use [context](https://github.com/thoughtbot/shoulda-context) but that would add another dependency to the project. 

If you run the tests you'll see a lot of `warning: instance variable @doc not initialized`. I followed the pattern for @name, which also issues this warning (in the master branch). I'm down for a follow up PR to address that, and those `warning: File.exists? is a deprecated name, use File.exist? instead` deprecation warnings. But I figured that was out of scope for AVRO-2001. 

Thanks again for your prompt feedback. It will make my task a lot easier to have this supported in the core gem!

tjwp commented on lang/ruby/lib/avro/schema.rb:
```diff
@@ -199,12 +202,12 @@ def self.make_field_objects(field_data, names, namespace=nil)
       end
 
-      def initialize(name, namespace, fields, names=nil, schema_type=:record)
+      def initialize(name, namespace, fields, names=nil, schema_type=:record, doc=nil)
```
I think there should be an `attr_reader :doc` for `RecordSchema`.

tjwp commented on lang/ruby/lib/avro/schema.rb:
```diff
@@ -285,12 +286,12 @@ def to_avro(names=Set.new)
 
     class EnumSchema < NamedSchema
       attr_reader :symbols
```
I think there should be an `attr_reader :doc` for `EnumSchema`.

tjwp commented on lang/ruby/test/test_schema.rb:
```diff
@@ -160,4 +160,101 @@ def test_to_avro_handles_falsey_defaults
+
+    assert_equal field_schema_hash, field_schema_json.to_avro
+
+    # Testing doc support for records
```
I'd make these separate test methods: `test_field_doc_attribute`, `test_record_doc_attribute`, etc. I don't think it adds any value to combine them. If one breaks, then they're different areas of the code.

tjwp commented on lang/ruby/test/test_protocol.rb:
```diff
@@ -196,4 +219,12 @@ def test_namespaces
+    # Testing doc support at the protocol level
+    assert_equal 'documentation', original.doc
+    # Testing doc support at the messages level
+    assert_equal 'documentation', original.messages['hello'].doc
```
This is kind of nitpicky, but I'd probably use different strings like `'protocol_doc'` and `'message_doc'` for the two different doc values. Otherwise, technically you could make a mistake in the implementation and pull the doc from the same place for both assertions.

lostphilosopher commented:
@tjwp Solid feedback again; good catch on the missing `attr_reader`s! I don't mind "nit picks" at all - I'd like to get this right. I've pushed up my code with your recommended changes, and rebased back into 1 commit. Thank you for your help with this, much appreciated!

lostphilosopher commented:
@tjwp this PR has been sitting out here awhile, is there something I can or should do get this merged? My team is using a fork on prod right now, which works, but we'd rather not do that indefinitely...

(I realize you don't merge PRs, but I figured you'd know the process and maybe the likely timeline?)

Thanks!

tjwp commented:
@lostphilosopher I'm in the same boat with [AVRO-1695](https://github.com/apache/avro/pull/116) and [AVRO-1969](https://github.com/apache/avro/pull/170), so I don't really have any answers for you.

lostphilosopher commented:
@rdblue @busbey @spacharya hi folks, sorry to "cold mention" you. You all seem active in the Ruby side of this project so I decided to reach out. This is my first PR to this project and I'm curious how long it will take to get this merged? Did I miss any steps in the process? If the answer is that this could take a long time that's fine with me. I'd just like to know; and to be sure the hold up isn't on my end. Thank you all.

spacharya commented:
This looks good to me.
I see other peoples approvals too :).
If no one has any objection then I will merge this into master in the next week.

lostphilosopher commented:
@spacharya awesome! Thank you!

busbey commented:
@tjwp reviews do help a ton. Do you have a preferred email address to have published as a reviewer of this change? Any other things you've reviewed that you'd like to point me towards as I work through a round of Avro changes?

spacharya commented:
Hi,
I have committed this to master.
However, I forgot to close the commit as I was pushing it. @lostphilosopher  can you close this pull request. I will make my life a lot easier.
Thanks.

lostphilosopher commented:
@spacharya consider it done!

lostphilosopher commented:
Thank you very much @spacharya, @busbey, and @tjwp, this makes my life easier - and makes my day! 😄

tjwp commented:
@busbey My preferred email would be (tjwp @ cantab.net). Thanks!

I don't have anything else to point you to right now. I try to look at any Ruby changes that go by since that is the implementation that I use primarily.
